### PR TITLE
Link reify

### PR DESCRIPTION
**Pull request by krl**

### CODE REVIEW

File: src/link.rs [hunk 1]
```diff
@@ -141,41 +142,77 @@ where
...
+                            drop(borrow);
+                            let borrow = self.inner.borrow();
+
+                            Ok(LinkCompound(borrow))
+                        } else {
+                            unreachable!()
+                        }
+                    }
+                    Err(CanonError::NotFound) => {
+                        println!("not found in store");
```
Comment (ZER0) [inline src/link.rs:168]: Remove the `println`.

File: src/link.rs [hunk 2]
```diff
@@ -141,41 +142,77 @@ where
...
+                                        Ok(LinkCompound(borrow))
+                                    }
+                                    Err(PersistError::Canon(e)) => Err(e),
+                                    err
+                                    @
+                                    Err(
+                                        PersistError::Io(_)
+                                        | PersistError::Other(_),
+                                    ) => {
+                                        println!("persist error: {:?}", err);
```
Comment (ZER0) [inline src/link.rs:202]: Remove the `println`

File: src/link.rs [hunk 3]
```diff
@@ -141,41 +142,77 @@ where
...
-
-                                Ok(LinkCompound(borrow))
-                            }
-                            Err(PersistError::Canon(e)) => Err(e),
-                            Err(
-                                PersistError::Io(_) | PersistError::Other(_),
-                            ) => {
-                                // TODO: log errors to the backend
-                                Err(CanonError::NotFound)
+            LinkInner::Ia(id, _) => {
```
Comment (ZER0) [inline src/link.rs:145]: The whole branch that is following contains a lot of if / match nested, some code duplication, and it's not very readable. Could you try to do a bit of refactoring here? In some cases I think you can take advantages of guards for example. reducing the nesting.
Also, a bit more of documentation (comments) would be nice.

File: tests/persistence.rs [hunk 4]
```diff
@@ -74,41 +73,26 @@ mod persist_tests {
...
-            assert_eq!(list.pop()?, Some(n - i - 1));
-        }
-
-        // then the restored copy
-
-        for i in 0..n {
-            assert_eq!(restored.pop()?, Some(n - i - 1));
-        }
+        persist()
+    }
```
Comment (ZER0) [inline tests/persistence.rs:84]: You should add a comment here where you explain the reason of those tests, and that we rely on the env to run the tests in threads (e.g. you can explicitly say that having `RUST_TEST_THREADS` set to `1` in the env would make the tests useless, they need multi threads)

**Review by ZER0 — CHANGES_REQUESTED** [review]
Overall looks good, there are few nits and I will try to refactor a bit the code to have less nesting if possible, and more documentation / comments.

Also, ensure you add the related issue to both the PR and the commit message (it's the #68)

File: src/link.rs (hunk 3 above)
Comment (ZER0) [reply src/link.rs:145]: I still believe there is too much nesting, but the comments help. We can land as is and trying to improve later.

**Review by ZER0 — APPROVED** [review]
LGTM, but please add the comment I mentioned here: https://github.com/dusk-network/microkelvin/pull/69/files#r668748673

File: tests/persistence.rs (hunk 4 above)
Comment (krl) [reply tests/persistence.rs:84]: it's already there, above

**Pull request merged**